src/app/clusters/bindings/BindingManager.h

```diff
@@ -117,6 +153,7 @@ class BindingManager
 
     PendingNotificationMap mPendingNotificationMap;
     BoundDeviceChangedHandler mBoundDeviceChangedHandler;
+    BoundDeviceContextReleaseHandler mBoundDeviceContextReleaseHandler;
```

Comment:
I think we should move this member into the pending notification map.  Here's why: there are lots of things that can remove an entry from the pending notification map and that are not correctly decrementing the refcount and calling the release handler with this PR.  Some examples:
* The `RemoveAllEntriesForNode` call in `BindingManager::EstablishConnection`.
* The `RemoveAllEntriesForFabric` call in `BindingManager::FabricRemoved`
* The `RemoveEntry` call in `BindingManager::UnicastBindingRemoved`

This would be easier to ger right if the map itself were calling the callback and managing the state appropriately whenever things are removed from it.

Reply:
Fair point. I introduced several changes with regards to that:
* I renamed `BindingManagerContext` to `PendingNotificationContext` and moved to the `PendingNotificationmMap`
* I moved release context handler object from `BingingManager` to the `PendingNotificationmMap`
* Added to the `PendingNotificationContext` functionality that automatically calls release context handler if customers counter reaches 0.
* Added allocating `PendingNotificationContext` in the `PendingNotificationmMap` method.
* Added decrementing and incrementing customer counter in remove and add notification methods.